src/back.rs

```diff
@@ -4,111 +4,140 @@
 
 use std::boxed::FnBox;
 use std::collections::HashMap;
-use std::sync::mpsc;
+use std::sync::Mutex;
 use std::thread;
-use {Handle, Request, Wait};
+use pulse::{Pulse, Trigger, Select, Barrier};
+
+use {Handle, Wait};
 
 
 struct Pending {
     task: Box<FnBox() + Send>,
-    dependencies: Vec<Handle>,
+    done: Trigger
 }
 
-struct Active {
-    link: thread::JoinHandle<()>,
+struct Inner {
+    exit: Option<Trigger>,
+    exit_method: Wait,
+
+    pending_select: Select,
+    pending: HashMap<usize, Pending>,
+
+    active_select: Select,
+    active: HashMap<usize, thread::JoinHandle<()>>
 }
 
 /// Task queue back-end.
 pub struct Backend {
-    output: mpsc::Sender<Request>,
-    pending: HashMap<Handle, Pending>,
-    active: HashMap<Handle, Active>,
+    inner: Mutex<Inner>
```

Comment:
Does it make sense to have `Backend` consisting just of a single field? I think it would be more sound to rename `Index` -> `Backend`, and remove this trivial existing backend type.